File: src/main/java/nieboczek/lifestolen/serializer/lang/SerializedStringBuilder.java
package nieboczek.lifestolen.serializer.lang;

public final class SerializedStringBuilder {
    private final StringBuilder builder;
    private int indentLevel = 0;

    public SerializedStringBuilder() {
        builder = new StringBuilder();
    }

    public void indent() {
        indentLevel++;
    }

    public void unindent() {
        if (indentLevel == 0)
            throw new IllegalStateException("Tried to decrease the indent level to a negative value");

        indentLevel--;
    }

    public SerializedStringBuilder indented() {
        builder.append("  ".repeat(indentLevel));
        return this;
    }

    public SerializedStringBuilder text(String text) {
        builder.append(text);
        return this;
    }

    public SerializedStringBuilder text(char text) {
        builder.append(text);
        return this;
    }

    public SerializedStringBuilder newLine() {
        builder.append('\n');
        return this;
    }

    public String getString() {
        return builder.toString();
    }
}

File: src/main/kotlin/nieboczek/lifestolen/serializer/lang/SerializedStringBuilder.kt
package nieboczek.lifestolen.serializer.lang

class SerializedStringBuilder {
    private val builder = StringBuilder()
    private var indentLevel = 0

    val string: String
        get() = builder.toString()

    fun indent() {
        indentLevel++
    }

    fun unindent() {
        check(indentLevel != 0) { "Tried to decrease the indent level to a negative value" }
        indentLevel--
    }

    fun indented(): SerializedStringBuilder {
        builder.append("  ".repeat(indentLevel))
        return this
    }

    fun text(text: String): SerializedStringBuilder {
        builder.append(text)
        return this
    }

    fun text(text: Char): SerializedStringBuilder {
        builder.append(text)
        return this
    }

    fun newLine(): SerializedStringBuilder {
        builder.append('\n')
        return this
    }
}
